src/meta-srv/src/procedure/region_migration/downgrade_leader_region.rs

```diff
@@ -13,23 +13,461 @@
+    /// - [MailboxTimeout](error::Error::MailboxTimeout), Timeout.
+    /// - Failed to downgrade region on the Datanode.
+    /// - [MailboxReceiver](error::Error::MailboxReceiver), The sender is dropped without sending (impossible).
+    /// - [UnexpectedInstructionReply](error::Error::UnexpectedInstructionReply).
+    /// - Invalid JSON.
+    async fn downgrade_region(
+        &self,
+        ctx: &mut Context,
+        downgrade_instruction: &Instruction,
+    ) -> Result<()> {
+        let pc = &ctx.persistent_ctx;
+        let region_id = pc.region_id;
+        let leader = &pc.from_peer;
+
+        let msg = MailboxMessage::json_message(
+            &format!("Downgrade leader region: {}", region_id),
+            &format!("Meta@{}", ctx.server_addr()),
+            &format!("Datanode-{}@{}", leader.id, leader.addr),
+            common_time::util::current_time_millis(),
+            downgrade_instruction,
+        )
+        .with_context(|_| error::SerializeToJsonSnafu {
+            input: downgrade_instruction.to_string(),
+        })?;
+
+        let ch = Channel::Datanode(leader.id);
+        let receiver = ctx
+            .mailbox
+            .send(&ch, msg, DOWNGRADE_LEADER_REGION_TIMEOUT)
+            .await?;
+
+        let msg = receiver.await??;
+        let reply = HeartbeatMailbox::json_reply(&msg)?;
+        let InstructionReply::DowngradeRegion(DowngradeRegionReply {
+            last_entry_id,
+            exist,
+            error,
+        }) = reply
+        else {
+            return error::UnexpectedInstructionReplySnafu {
+                mailbox_message: msg.to_string(),
+                reason: "expect downgrade region reply",
+            }
+            .fail();
+        };
+
+        if error.is_some() {
+            return error::UnexpectedSnafu {
+                violated: format!(
+                    "Failed to downgrade the region {} on Datanode {:?}, error: {:?}",
+                    region_id, leader, error
+                ),
+            }
+            .fail();
+        }
+
+        if !exist {
+            warn!(
+                "Trying to downgrade the region {} on Datanode {}, but region doesn't exist!",
+                region_id, leader
+            );
+        }
+
+        if let Some(last_entry_id) = last_entry_id {
+            ctx.volatile_ctx.set_last_entry_id(last_entry_id);
+        }
+
+        Ok(())
+    }
+
+    /// Downgrades a leader region.
+    ///
+    /// Fast path:
+    /// - Waits for the reply of downgrade instruction.
+    ///
+    /// Slow path:
+    /// - Waits for the lease of the leader region expired.
+    async fn downgrade_region_with_retry(&self, ctx: &mut Context) {
+        let instruction = self.build_downgrade_region_instruction(ctx);
+
+        let mut retry = 0;
+
+        while retry < self.optimistic_retry {
+            if let Err(err) = self.downgrade_region(ctx, &instruction).await {
```

Comment:
Better to distinguish the retryable errors.

Reply:
Resolved